File: src/main/java/com/github/ofrostdev/api/utils/ConcurrentCache.java
package com.github.ofrostdev.api.utils;

import com.google.common.collect.ImmutableMap;
import java.util.Collection;
import java.util.Iterator;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public abstract class ConcurrentCache<K, V> {
    private Map<K, V> elements = new ConcurrentHashMap<>();

    public Map<K, V> getElements() {
        return this.elements;
    }

    public boolean contains(K element) {
        return (this.elements.get(element) != null);
    }

    public void addElement(K key, V value) {
        this.elements.put(key, value);
    }

    public void removeElement(K toRemove) {
        this.elements.remove(toRemove);
    }

    public V get(K key) {
        return this.elements.get(key);
    }

    public Collection<V> getAll() {
        return getElements().values();
    }

    public Optional<V> find(K key) {
        return Optional.ofNullable(get(key));
    }

    public ImmutableMap<K, V> toImmutable() {
        return ImmutableMap.copyOf(this.elements);
    }

    public Iterator<V> iterator() {
        return this.elements.values().iterator();
    }

    public Iterator<K> iteratorKeySet() {
        return this.elements.keySet().iterator();
    }

    public int size() {
        return this.elements.size();
    }
}

File: src/main/kotlin/com/github/ofrostdev/api/utils/common/ConcurrentCache.kt
package com.github.ofrostdev.api.utils.common

import com.google.common.collect.ImmutableMap
import java.util.*
import java.util.concurrent.ConcurrentHashMap

abstract class ConcurrentCache<K, V> {
    private val elements: MutableMap<K, V> = ConcurrentHashMap()

    fun getElements(): Map<K, V> {
        return this.elements
    }

    fun contains(element: K): Boolean {
        return (elements[element] != null)
    }

    fun addElement(key: K, value: V) {
        elements[key] = value
    }

    fun removeElement(toRemove: K) {
        elements.remove(toRemove)
    }

    fun get(key: K): V? {
        return elements[key]
    }

    val all: Collection<V>
        get() = getElements().values

    fun find(key: K): Optional<V & Any> {
        return Optional.ofNullable(get(key))
    }

    fun toImmutable(): ImmutableMap<K, V> {
        return ImmutableMap.copyOf(this.elements)
    }

    fun iterator(): Iterator<V> {
        return elements.values.iterator()
    }

    fun iteratorKeySet(): Iterator<K> {
        return elements.keys.iterator()
    }

    fun size(): Int {
        return elements.size
    }
}